sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala

```diff
@@ -388,6 +388,24 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
     assert(df.queryExecution.executedPlan.collect { case e: Exchange => e }.isEmpty)
   }
 
+  test("SPARK-44505: should not call planInputPartitions() on explain") {
+    val df = spark.read.format(classOf[ScanDefinedColumnarSupport].getName)
+      .option("columnar", "PARTITION_DEFINED").load()
+    // Default mode will throw an exception on explain.
+    intercept[IllegalArgumentException](df.explain())
```

Comment:
I think it would be better to add assertions for the exception content, if possible.

Reply:
Done.